### PR TITLE
Make `transaction view` emit JSON

### PR DESCRIPTION
**Pull request by carlhammann**
# Changelog

```yaml
- description: |
    Make `cardano-cli * transaction view` output JSON by default.
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

We discussed [here](https://github.com/input-output-hk/cardano-cli/pull/303#discussion_r1335088169) that `cardano-cli` should output JSON by default. The `transaction view` commands currently write YAML. This PR makes it so that YAML is only emitted by these commands with the option `--output-format yaml`. I also added an `--out-file` option to redirect the output to some file instead of stdout.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Transaction.hs [hunk 1]
```diff
@@ -276,4 +276,4 @@ pTransactionId  :: Parser (TransactionCmds era)
 pTransactionId = TxGetTxId <$> pInputTxOrTxBodyFile
 
 pTransactionView :: Parser (TransactionCmds era)
-pTransactionView = TxView <$> pInputTxOrTxBodyFile
+pTransactionView = TxView <$> pTxViewOutputFormat <*> pMaybeOutputFile <*> pInputTxOrTxBodyFile
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Transaction.hs]: Please format like this:

```
pTransactionView =
  TxView
    <$> pTxViewOutputFormat
    <*> pMaybeOutputFile
    <*> pInputTxOrTxBodyFile
```

File: cardano-cli/src/Cardano/CLI/Legacy/Options.hs [hunk 2]
```diff
@@ -668,7 +668,7 @@ pTransaction envCli =
   pTransactionId = TxGetTxId <$> pInputTxOrTxBodyFile
 
   pTransactionView :: Parser LegacyTransactionCmds
-  pTransactionView = TxView <$> pInputTxOrTxBodyFile
+  pTransactionView = TxView <$> pTxViewOutputFormat <*> pMaybeOutputFile <*> pInputTxOrTxBodyFile
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Legacy/Options.hs]: Please format

**Review by newhoggy — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/Json/Friendly.hs [hunk 3]
```diff
@@ -43,12 +42,58 @@ import           Data.Yaml.Pretty (setConfCompare)
...
 import           GHC.Real (denominator)
 import           GHC.Unicode (isAlphaNum)
+import qualified Data.Aeson.Encode.Pretty as Aeson
+import Control.Monad.Trans (MonadIO)
+
+jsonConfig :: Aeson.Config
+jsonConfig = Aeson.defConfig {Aeson.confCompare = compare}
+
+friendlyTxJson ::
+  (MonadIO m, IsCardanoEra era)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Json/Friendly.hs]:
```suggestion
  (MonadIO m)
```
Redundant constraint. `IsCardanoEra era` is obtainable using `cardanoEraConstraints` function.

File: cardano-cli/src/Cardano/CLI/Json/Friendly.hs [hunk 4]
```diff
@@ -43,12 +42,58 @@ import           Data.Yaml.Pretty (setConfCompare)
...
+  -> CardanoEra era
+  -> Tx era
+  -> m (Either (FileError e) ())
+friendlyTxJson mOutFile era = writeLazyByteStringOutput mOutFile .
+  Aeson.encodePretty' jsonConfig .
+  object .
+  friendlyTx era
+
+friendlyTxBodyJson ::
+  (MonadIO m, IsCardanoEra era)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Json/Friendly.hs]:
```suggestion
  (MonadIO m)
```

File: cardano-cli/src/Cardano/CLI/Json/Friendly.hs [hunk 5]
```diff
@@ -43,12 +42,58 @@ import           Data.Yaml.Pretty (setConfCompare)
...
+  object .
+  friendlyTxBody era
 
 yamlConfig :: Yaml.Config
 yamlConfig = Yaml.defConfig & setConfCompare compare
 
-friendlyTxBS :: IsCardanoEra era => CardanoEra era -> Tx era -> ByteString
-friendlyTxBS era = Yaml.encodePretty yamlConfig . object . friendlyTx era
+friendlyTxYaml ::
+  (MonadIO m, IsCardanoEra era)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Json/Friendly.hs]:
```suggestion
  (MonadIO m)
```

File: cardano-cli/src/Cardano/CLI/Json/Friendly.hs [hunk 6]
```diff
@@ -43,12 +42,58 @@ import           Data.Yaml.Pretty (setConfCompare)
...
+  -> CardanoEra era
+  -> Tx era
+  -> m (Either (FileError e) ())
+friendlyTxYaml mOutFile era = writeByteStringOutput mOutFile .
+  Yaml.encodePretty yamlConfig .
+  object .
+  friendlyTx era
+
+friendlyTxBodyYaml ::
+  (MonadIO m, IsCardanoEra era)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Json/Friendly.hs]:
```suggestion
  (MonadIO m)
```

**Review by carbolymer — APPROVED** [review]

**Pull request merged**